File: 4.8-bc/ts/stage-3-hexagonal/src/commerce/app/service.ts
// Commerce use case — створення замовлення.
import { randomUUID } from "node:crypto";

import type { EventBus } from "../../shared/events/bus.js";
import { InvalidOrderError } from "../domain/errors.js";
import { OrderPlaced, type Order } from "../domain/order.js";
import type { Repository } from "../domain/repository.js";

export class Service {
  constructor(
    private readonly repo: Repository,
    private readonly bus: EventBus,
  ) {}

  async place(userId: string, totalCents: number): Promise<Order> {
    if (totalCents < 0) throw new InvalidOrderError();

    const o: Order = {
      id: randomUUID(),
      userId,
      totalCents,
      status: "pending",
      createdAt: new Date(),
    };
    await this.repo.create(o);
    await this.bus.publish(
      new OrderPlaced(o.id, o.userId, o.totalCents, o.createdAt),
    );
    return o;
  }
}

File: 4.8-bc/go/stage-2-feature-first/app/commerce/service.go
// Package commerce — Commerce feature: замовлення.
package commerce

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type Service struct {
	repo *Repository
}

func NewService(repo *Repository) *Service {
	return &Service{repo: repo}
}

func (s *Service) Place(ctx context.Context, userID uuid.UUID, totalCents int64) (*Order, error) {
	o := Order{
		ID:         uuid.New(),
		UserID:     userID,
		TotalCents: totalCents,
		Status:     "pending",
		CreatedAt:  time.Now().UTC(),
	}
	if err := s.repo.Create(ctx, o); err != nil {
		return nil, err
	}
	return &o, nil
}
